compiler/src/dotty/tools/dotc/transform/sjs/PrepJSInterop.scala

```diff
@@ -0,0 +1,1221 @@
+          if (shouldPrepareExports)
+            registerClassOrModuleExports(sym)
+
+          if (isJSAny(sym))
+            transformJSClassDef(tree)
+          else
+            transformScalaClassDef(tree)
+
+        case _: TypeDef =>
+          super.transform(tree)
+
+        case tree: ValOrDefDef =>
+          /* Prepare exports for methods, local defs and local variables.
+           * Avoid *fields* (non-local non-method) because they all have a
+           * corresponding getter, which is the one that handles exports.
+           * (Note that local-to-block can never have exports, but the error
+           * messages for that are handled by genExportMember).
+           */
+          if (shouldPrepareExports && (sym.is(Method) || sym.isLocalToBlock)) {
+            exporters.getOrElseUpdate(sym.owner, mutable.ListBuffer.empty) ++=
+              genExportMember(sym)
+          }
+
+          if (sym.isLocalToBlock) {
+            super.transform(tree)
+          } else if (isJSNative) {
+            transformJSNativeValOrDefDef(tree)
+          } else if (enclosingOwner is OwnerKind.JSType) {
+            transformValOrDefDefInJSType(tree)
+          } else {
+            transformScalaValOrDefDef(tree)
+          }
+      }
+
+      /* Give tree.symbol, not sym, so that for modules it is the module
+       * symbol, not the module class symbol.
+       *
+       * #1899 This must be done *after* transforming the member def tree,
+       * because fixPublicBeforeTyper must have run.
+       */
+      markExposedIfRequired(tree.symbol)
+
+      transformedTree
+    }
+
+    private def transformScalaClassDef(tree: TypeDef)(using Context): Tree = {
+      val sym = tree.symbol
+
+      // In native JS things, only js.Any stuff is allowed
+      if (enclosingOwner is OwnerKind.JSNative) {
+        /* We have to allow synthetic companion objects here, as they get
+         * generated when a nested native JS class has default arguments in
+         * its constructor (see #1891).
+         */
+        if (!sym.is(Synthetic)) {
+          report.error(
+              "Native JS traits, classes and objects cannot contain inner " +
+              "Scala traits, classes or objects (i.e., not extending js.Any)",
+              tree)
+        }
+      }
+
+      if (sym == jsdefn.PseudoUnionClass)
+        sym.addAnnotation(jsdefn.JSTypeAnnot)
+
+      val kind =
+        if (sym.is(Module)) OwnerKind.ScalaMod
+        else OwnerKind.ScalaClass
+      enterOwner(kind) {
+        super.transform(tree)
+      }
+    }
+
+    private def transformScalaValOrDefDef(tree: ValOrDefDef)(using Context): Tree = {
+      // There is nothing special to do for a Scala val or def
+      super.transform(tree)
+    }
+
+    private def transformTemplate(tree: Template)(using Context): Template = {
+      // First, recursively transform the template
+      val transformedTree = super.transform(tree).asInstanceOf[Template]
```

Comment:
In dotc it's not an issue, by construction, because self-types are not transformed through the regular `transform`. They are rerouted to `transformSelf` by `MacroTransfrom.Transformer`. Since we do not override `transformSelf`, we have nothing special to do.